File: apex-forex-bot/backtest.py
"""Backtest pe STRATEGIA REALĂ — rubrica AI calculată mecanic + filtrele +
exiturile live din apex/position.py (același cod care tranzacționează).

Rulare:
    python backtest.py                          # OANDA/TD cu cheile din .env
    BT_SYMBOL=GBP_USD python backtest.py
    BT_SYNTHETIC=true python backtest.py        # fără internet — validare motor

Ce simulează fidel: criteriile de intrare din promptul AI, MIN_CRITERIA,
filtrul de trend 1h, veto-ul Livermore+Turtle, cooldown după pierdere, sizing
pe risc 2%, spread plătit la intrare, slippage, exituri identice cu live-ul
(SL/TP/breakeven/trailing/runner) evaluate la close — cum face botul la tick.
Ce NU simulează: judecata LLM-ului (filtru suplimentar la live → live-ul ia de
regulă MAI PUȚINE trade-uri). Rezultatele trecute nu garantează profit.
"""
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
os.environ.setdefault("PAPER_TRADING", "true")

from apex import config as cfg, forex, indicators, strategies, logger  # noqa: E402
from apex import position as position_mod  # noqa: E402
from apex.position import calc_entry_sltp, trail_stop, exit_trigger  # noqa: E402

SYMBOL = os.getenv("BT_SYMBOL") or cfg.SYMBOL
START_BAL = float(os.getenv("BT_BALANCE") or 1000)
CANDLES = int(os.getenv("BT_CANDLES") or 2000)
SLIPPAGE_PIPS = float(os.getenv("BT_SLIPPAGE_PIPS") or 0.3)
SPREAD_PIPS = float(os.getenv("BT_SPREAD_PIPS") or 1.0)
SYNTHETIC = os.getenv("BT_SYNTHETIC") == "true"
MIN_CRITERIA = int(os.getenv("MIN_CRITERIA") or 4)
# Per-user settings live in Redis, not cfg — the live loop reads them off the
# user record, so they have to be injectable here or the backtest silently
# simulates a different account than the one trading (e.g. cfg risk 1.25% vs
# the 35% actually configured on the forex bot).
RISK = float(os.getenv("BT_RISK") or cfg.RISK_PER_TRADE)
TRAILING = (os.getenv("BT_TRAILING") or "true") != "false"
BREAKEVEN_R = float(os.getenv("BT_BREAKEVEN_R") or 0)
HTF_ON = (os.getenv("BT_HTF") or ("true" if cfg.HTF_FILTER else "false")) != "false"
# Trail one CONSTANT risk-unit behind price instead of live's recomputed
# abs(entry - cur_sl). See position.trail_stop() — off by default so the
# baseline run stays a faithful mirror of what trades today.
FIXED_R_TRAIL = os.getenv("BT_FIXED_R_TRAIL") == "true"
# BT_STRATEGY: criteria (rubrica AI istorică) | mean_reversion | trend | breakout
BT_STRATEGY = (os.getenv("BT_STRATEGY") or "criteria").lower().replace("mean", "mean_reversion") if (os.getenv("BT_STRATEGY") or "criteria").lower() == "mean" else (os.getenv("BT_STRATEGY") or "criteria").lower()  # 4/5 — forex: mc4>mc5 (mc5 prea puține semnale pe 3000 lumânări)

logger.info = lambda *a, **k: None  # fără spam pe mii de lumânări


def criteria_signal(ind, strat):
    """Rubrica din promptul AI (apex/ai.py), calculată mecanic."""
    rsi, macd_h = float(ind["rsi"]), float(ind["macdHist"])
    price, ema20 = float(ind["price"]), float(ind["ema20"])
    has_vol = ind.get("hasVolume", True)
    vol_ok = float(ind["volumeRatio"]) > 1.2 if has_vol else None
    srsi_k = float(ind["stochRsiK"])

    def score(direction):
        s = 0
        bull = direction == "BUY"
        if ind["emaTrend"] == ("BULLISH" if bull else "BEARISH"):
            s += 1
        if (rsi < 50 if bull else rsi > 50) or ind["divergence"] == ("BULLISH" if bull else "BEARISH"):
            s += 1
        if (macd_h > 0) == bull:
            s += 1
        # criteriul de volum — la fel ca ai.py: fără volum → Stoch RSI aliniat
        if vol_ok if vol_ok is not None else ((srsi_k < 50) == bull):
            s += 1
        if (price < ema20) == bull:
            s += 1
        liv, tur, sor = strat["livermore"], strat["turtle"], strat["soros"]
        if ((tur.get("breakoutStr") == "STRONG" and tur.get("signal") == direction)
                or (liv["trend"] == ("BULLISH" if bull else "BEARISH") and liv.get("strength", 0) >= 0.5)
                or sor["direction"] == ("BULLISH" if bull else "BEARISH")):
            s += 1
        return min(5, s)

    b, s = score("BUY"), score("SELL")
    if b >= MIN_CRITERIA and b > s:
        return {"action": "BUY", "criteriaScore": b}
    if s >= MIN_CRITERIA and s > b:
        return {"action": "SELL", "criteriaScore": s}
    return {"action": "HOLD", "criteriaScore": 0}


def resample_1h(candles, ratio=12):
    out = []
    for i in range(0, len(candles) - ratio + 1, ratio):
        grp = candles[i:i + ratio]
        out.append({"time": grp[0]["time"], "open": grp[0]["open"],
                    "close": grp[-1]["close"],
                    "high": max(c["high"] for c in grp),
                    "low": min(c["low"] for c in grp),
                    "volume": sum(c["volume"] for c in grp)})
    return out


def synthetic_candles(n):
    seed = [42]

    def rnd():
        seed[0] = (seed[0] * 1103515245 + 12345) % 2 ** 31
        return seed[0] / 2 ** 31

    out, p, drift = [], 1.0850, 0.0
    for i in range(n):
        if i % 250 == 0:
            drift = (rnd() - 0.5) * 0.00006
        ret = drift + (rnd() - 0.5) * 0.0004
        o, p = p, max(0.5, p * (1 + ret))
        out.append({"time": i, "open": o, "close": p,
                    "high": max(o, p) * (1 + rnd() * 0.0001),
                    "low": min(o, p) * (1 - rnd() * 0.0001),
                    "volume": 0})
    return out


def fetch_candles():
    if SYNTHETIC:
        return synthetic_candles(CANDLES + 300)
    from apex.brokers import get_broker
    broker = get_broker()
    return broker.get_candles(SYMBOL, cfg.TIMEFRAME, min(CANDLES + 300, 5000))


def run():
    pip = forex.pip_size(SYMBOL)
    half_spread = SPREAD_PIPS / 2 * pip
    slip = SLIPPAGE_PIPS * pip
    print("\n" + "═" * 64)
    print(f"  📊 APEX FOREX BACKTEST — metoda: {BT_STRATEGY} (exituri user_loop)")
    print(f"  {'⚠️  DATE SINTETICE — validare motor, NU concluzii de profit' if SYNTHETIC else f'Symbol: {SYMBOL} | TF: {cfg.TIMEFRAME}'}")
    print(f"  Balanță: ${START_BAL:.0f} | SL {position_mod.USERLOOP_SL_ATR_MULT:g}×ATR / "
          f"TP {position_mod.USERLOOP_TP_ATR_MULT:g}×ATR | risc {RISK * 100:g}% (pe startBalance) | "
          f"spread {SPREAD_PIPS:g}p | slippage {SLIPPAGE_PIPS:g}p")
    print(f"  Trailing: {'1R' if TRAILING else 'off'} | breakeven: "
          f"{f'{BREAKEVEN_R:g}R' if BREAKEVEN_R else 'off'} | HTF: {'on' if HTF_ON else 'off'} | "
          f"exit SL/TP intrabar (high/low)"
          f"{' | trail=1R FIX' if FIXED_R_TRAIL else ''}")
    print("═" * 64)

    candles = fetch_candles()
    if len(candles) < 400:
        raise SystemExit(f"Doar {len(candles)} lumânări — minim 400")

    balance, position, trades = START_BAL, None, []
    last_loss_idx = -10 ** 9
    # Live blocks re-entry for _LOSS_COOLDOWN_MIN (15) after a losing close,
    # not cfg.COOLDOWN_AFTER_LOSS_MIN (3 in scalp mode) — another silent gap.
    cooldown_bars = max(1, int(os.getenv("BT_COOLDOWN_MIN") or 15) // 5)
    peak, max_dd, spread_cost = START_BAL, 0.0, 0.0

    def close(exit_mid, reason, i):
        nonlocal balance, position, last_loss_idx, peak, max_dd, spread_cost
        d = 1 if position["side"] == "BUY" else -1
        exit_px = exit_mid - d * (half_spread + slip)
        pnl = forex.pnl_usd(position["side"], position["entryPrice"], exit_px,
                            position["quantity"], SYMBOL)
        spread_cost += (half_spread + slip) * position["quantity"] * (1 if SYMBOL.endswith("_USD") else 1 / exit_mid)
        balance += pnl
        pips = forex.to_pips(
            (exit_px - position["entryPrice"]) * d, SYMBOL, exit_px)
        trades.append({"side": position["side"], "pnl": pnl, "reason": reason,
                       "pips": round(pips, 1)})
        if pnl < 0:
            last_loss_idx = i
        position = None
        peak = max(peak, balance)
        max_dd = max(max_dd, (peak - balance) / peak)

    for i in range(300, len(candles)):
        window = candles[:i + 1]
        bar = candles[i]
        mid = bar["close"]

        if position:
            # Order matters and mirrors live: cTrader holds SL/TP server-side,
            # so they fill intrabar against THIS bar's range using the stop
            # that was already resting when the bar opened. Only after that
            # does the bot's own 5-min tick get to trail the stop, and it sees
            # the close — never the wick it would have needed to react to.
            trigger = exit_trigger(position["side"], position["stopLoss"],
                                   position["takeProfit"], bar["high"], bar["low"])
            if trigger:
                # Broker fills AT the resting order's level, not at the close.
                level = (position["stopLoss"] if trigger == "STOP_LOSS"
                         else position["takeProfit"])
                close(level, trigger, i)
                continue
            moved = trail_stop(position["side"], position["entryPrice"],
                               position["stopLoss"], mid,
                               trailing=TRAILING, breakeven_r=BREAKEVEN_R,
                               initial_risk=(position["initialRisk"]
                                             if FIXED_R_TRAIL else None))
            if moved is not None:
                position["stopLoss"] = moved
        if position or balance < 10:
            continue
        if i - last_loss_idx < cooldown_bars:
            continue

        ind = indicators.analyze(window)
        strat = strategies.analyze(window)
        if BT_STRATEGY == "criteria":
            sig = criteria_signal(ind, strat)
        else:
            from apex import ai
            s2 = ai.signal_for_mode(BT_STRATEGY, ind, strat, None)
            sig = ({"action": s2["action"], "criteriaScore": s2.get("criteriaScore", 3)}
                   if s2["action"] in ("BUY", "SELL") and s2.get("confidence", 0) >= 62
                   else {"action": "HOLD", "criteriaScore": 0})
        if sig["action"] == "HOLD":
            continue

        liv, tur = strat["livermore"], strat["turtle"]
        contra = ("BUY" if liv["trend"] == "BEARISH" and tur.get("signal") == "SELL"
                  else "SELL" if liv["trend"] == "BULLISH" and tur.get("signal") == "BUY" else None)
        if liv.get("strength", 0) >= 0.8 and sig["action"] == contra:
            continue
        if HTF_ON:
            htf = strategies.htf_trend(resample_1h(window[-720:]))
            if os.getenv("BT_HTF_STRICT") == "true" or cfg.HTF_STRICT:
                # strict: intră DOAR pe direcția trendului 1h (NEUTRAL = HOLD)
                if htf != ("BULLISH" if sig["action"] == "BUY" else "BEARISH"):
                    continue
            elif ((sig["action"] == "BUY" and htf == "BEARISH")
                    or (sig["action"] == "SELL" and htf == "BULLISH")):
                continue

        d = 1 if sig["action"] == "BUY" else -1
        entry = mid + d * (half_spread + slip)  # intrarea plătește spread + slippage
        sl_px, tp_px = calc_entry_sltp(sig["action"], entry, float(ind["atr"]),
                                       SYMBOL, spread_pips=SPREAD_PIPS,
                                       sl_pips=cfg.STOP_LOSS_PIPS,
                                       tp_pips=cfg.TAKE_PROFIT_PIPS)
        stop_pips = forex.to_pips(abs(entry - sl_px), SYMBOL)
        mult = strategies.druckenmiller_multiplier(70, sig["criteriaScore"], liv, tur)
        # Live sizes off startBalance, never the running balance — deliberately,
        # so the same setup keeps the same lot size instead of drifting with
        # P&L. Compounding off `balance` here (as this did) inflates results
        # against a bot that does not compound.
        units = forex.calc_units(START_BAL, RISK, stop_pips, SYMBOL, entry,
                                 leverage=cfg.LEVERAGE, mult=mult)
        floor = forex.safe_min_units(SYMBOL, balance, entry, cfg.LEVERAGE,
                                     cfg.MARGIN_CAP)
        if floor == 0:  # account can't margin even one micro-lot — live skips
            continue
        units = forex.round_units(max(units, floor), SYMBOL)
        if units <= 0:
            continue
        position = {"symbol": SYMBOL, "side": sig["action"], "entryPrice": entry,
                    "quantity": units, "stopLoss": sl_px, "takeProfit": tp_px,
                    "initialStop": sl_px,
                    "initialRisk": abs(entry - sl_px)}
    if position:
        close(candles[-1]["close"], "END", len(candles) - 1)

    wins = [t for t in trades if t["pnl"] > 0]
    losses = [t for t in trades if t["pnl"] <= 0]
    gw = sum(t["pnl"] for t in wins)
    gl = abs(sum(t["pnl"] for t in losses))
    days = (len(candles) - 300) * 5 / 1440
    print(f"\n  Perioadă: ~{days:.1f} zile | Trades: {len(trades)} (✅ {len(wins)} / ❌ {len(losses)})")
    if trades:
        print(f"  Win rate: {len(wins) / len(trades) * 100:.1f}% | "
              f"Profit factor: {'∞' if gl == 0 else f'{gw / gl:.2f}'}")
        print(f"  Rezultat net: {(balance - START_BAL) / START_BAL * 100:+.2f}% (${balance:.2f}) | "
              f"Max DD: -{max_dd * 100:.1f}%")
        print(f"  Expectancy: {sum(t['pnl'] for t in trades) / len(trades):+.2f} $/trade | "
              f"Cost spread+slippage: ~${spread_cost:.2f}")
        reasons = {}
        for t in trades:
            reasons[t["reason"]] = reasons.get(t["reason"], 0) + 1
        print("  Exituri: " + " | ".join(f"{r}×{n}" for r, n in reasons.items()))
    print("\n  ⚠️  Stratul AI nu e simulat (la live filtrează în plus). Rezultatele")
    print("      trecute nu garantează nimic. Paper trading înainte de live.")
    print("═" * 64 + "\n")


if __name__ == "__main__":
    run()

File: apex-trade-bot/src/backtest.js
// ─── Backtest pe STRATEGIA REALĂ (nu un proxy diferit) ───────────────────────
// Rulare:  node src/backtest.js
//          BT_SYMBOL=XRPUSDT BT_TIMEFRAME=5m BT_CANDLES=1000 node src/backtest.js
//          BT_SYNTHETIC=true node src/backtest.js   (fără internet — doar validare motor)
//
// Ce simulează FIDEL:  rubrica de criterii din promptul AI (calculată mecanic),
//   gate-urile MIN_CRITERIA/MIN_VOLUME_RATIO, filtrul de trend 1h, veto-ul
//   Livermore+Turtle, cooldown după pierdere, sizing Druckenmiller, și exiturile
//   REALE din src/position.js (SL/TP/breakeven/trailing/runner) evaluate la close
//   de lumânare — exact cum face botul live la fiecare tick. Comisioane 0.1%/parte
//   + slippage pe intrare/ieșire.
// Ce NU simulează:  judecata LLM-ului. AI-ul poate respinge trade-uri pe care
//   rubrica le-ar lua (e un filtru în plus), deci live-ul ia DE REGULĂ mai puține
//   trade-uri decât acest backtest. Rezultatele NU sunt o promisiune de profit.

require('dotenv').config();
const indicators = require('./indicators');
const strategies = require('./strategies');
const logger     = require('./logger');
const cfg        = require('./config');
const { calcSLTP, checkPosition } = require('./position');

const SYMBOL    = process.env.BT_SYMBOL    || cfg.SYMBOL;
const TIMEFRAME = process.env.BT_TIMEFRAME || cfg.TIMEFRAME;
const CANDLES   = parseInt(process.env.BT_CANDLES || '1000');
const START_BAL = parseFloat(process.env.BT_BALANCE || '100');
const SLIPPAGE  = parseFloat(process.env.BT_SLIPPAGE || '0.0005'); // 0.05%/parte
const SYNTHETIC = process.env.BT_SYNTHETIC === 'true';

logger.info = () => {}; // fără spam de breakeven/runner pe mii de lumânări

// ─── Rubrica de criterii — identică cu cea cerută AI-ului în prompt ──────────
function criteriaSignal(ind, strat) {
  const rsi = parseFloat(ind.rsi), macdH = parseFloat(ind.macdHist);
  const volR = parseFloat(ind.volumeRatio), price = parseFloat(ind.price);
  const ema20 = parseFloat(ind.ema20);
  const bonus = (dir) =>
    (strat.turtle.breakoutStr === 'STRONG' && strat.turtle.signal === dir) ||
    (strat.livermore.trend === (dir === 'BUY' ? 'BULLISH' : 'BEARISH') && strat.livermore.strength >= 0.5) ||
    (strat.soros.direction === (dir === 'BUY' ? 'BULLISH' : 'BEARISH')) ? 1 : 0;

  let buy = 0;
  if (ind.emaTrend === 'BULLISH') buy++;
  if (rsi < 50 || ind.divergence === 'BULLISH') buy++;
  if (macdH > 0) buy++;
  if (volR > 1.2) buy++;
  if (price < ema20) buy++;
  buy += bonus('BUY');

  let sell = 0;
  if (ind.emaTrend === 'BEARISH') sell++;
  if (rsi > 50 || ind.divergence === 'BEARISH') sell++;
  if (macdH < 0) sell++;
  if (volR > 1.2) sell++;
  if (price > ema20) sell++;
  sell += bonus('SELL');

  const minC = parseInt(process.env.MIN_CRITERIA || '5');
  if (buy >= minC && buy > sell) return { action: 'BUY', criteriaScore: Math.min(5, buy) };
  if (sell >= minC && sell > buy) return { action: 'SELL', criteriaScore: Math.min(5, sell) };
  return { action: 'HOLD', criteriaScore: 0 };
}

// 5m → 1h pentru filtrul HTF (12 lumânări de 5m = 1h)
function resample1h(candles, ratio = 12) {
  const out = [];
  for (let i = 0; i + ratio <= candles.length; i += ratio) {
    const grp = candles.slice(i, i + ratio);
    out.push({
      time: grp[0].time, open: grp[0].open, close: grp[grp.length - 1].close,
      high: Math.max(...grp.map(c => c.high)), low: Math.min(...grp.map(c => c.low)),
      volume: grp.reduce((a, c) => a + c.volume, 0),
    });
  }
  return out;
}

// Date sintetice cu seed — DOAR pentru validarea motorului, nu pentru concluzii
function syntheticCandles(n) {
  let seed = 42;
  const rnd = () => (seed = (seed * 1103515245 + 12345) % 2 ** 31) / 2 ** 31;
  const out = []; let p = 100, drift = 0;
  for (let i = 0; i < n; i++) {
    if (i % 200 === 0) drift = (rnd() - 0.5) * 0.0015; // regimuri de trend
    const ret = drift + (rnd() - 0.5) * 0.006;
    const o = p; p = Math.max(1, p * (1 + ret));
    out.push({ time: i, open: o, close: p,
               high: Math.max(o, p) * (1 + rnd() * 0.002),
               low: Math.min(o, p) * (1 - rnd() * 0.002),
               volume: 800 + rnd() * 600 });
  }
  return out;
}

// OKX public, paginat — singura sursă publică ne-geo-blocată care dă istoric lung
// (Binance răspunde 451 din cloud US/EU; OKX /candles max 300, /history-candles max 100)
async function fetchOKXPaginated(symbol, bar, total) {
  const axios = require('axios');
  const instId = symbol.replace('USDT', '-USDT');
  const opts = { headers: { 'User-Agent': 'ApexTradeBot/2.0' }, timeout: 10000 };
  const parse = rows => rows.map(k => ({
    time: parseInt(k[0]), open: parseFloat(k[1]), high: parseFloat(k[2]),
    low: parseFloat(k[3]), close: parseFloat(k[4]), volume: parseFloat(k[5]),
  }));
  const { data } = await axios.get('https://www.okx.com/api/v5/market/candles',
    { ...opts, params: { instId, bar, limit: 300 } });
  if (data.code !== '0') throw new Error('OKX: ' + data.msg);
  let candles = parse(data.data); // newest-first
  while (candles.length < total) {
    const oldest = candles[candles.length - 1].time;
    const { data: h } = await axios.get('https://www.okx.com/api/v5/market/history-candles',
      { ...opts, params: { instId, bar, after: oldest, limit: 100 } });
    if (h.code !== '0' || !h.data?.length) break;
    candles = candles.concat(parse(h.data));
    await new Promise(r => setTimeout(r, 250)); // rate limit OKX: 10 req/2s
  }
  return candles.reverse(); // vechi → nou, cum așteaptă indicatorii
}

async function fetchCandles() {
  if (SYNTHETIC) return syntheticCandles(CANDLES + 200);
  const want = CANDLES + 250;
  const exchange = require('./exchange');
  let candles = [];
  try { candles = await exchange.getCandles(SYMBOL, TIMEFRAME, Math.min(want, 1000)); } catch {}
  if (candles.length < Math.min(want, 600)) {
    console.log(`[DATA] Doar ${candles.length} lumânări de la exchange — paginez OKX pentru istoric complet...`);
    candles = await fetchOKXPaginated(SYMBOL, TIMEFRAME, want);
  }
  console.log(`[DATA] ${candles.length} lumânări reale încărcate`);
  return candles;
}

async function runBacktest() {
  const FEE = cfg.FEE_PCT;
  console.log('\n' + '═'.repeat(64));
  console.log('  📊 APEX BACKTEST — strategia reală (rubrică AI + filtre + exituri live)');
  console.log(`  ${SYNTHETIC ? '⚠️  DATE SINTETICE — validare motor, NU concluzii de profit' : `Symbol: ${SYMBOL} | TF: ${TIMEFRAME}`}`);
  console.log(`  Balanță: $${START_BAL} | SL ${cfg.STOP_LOSS_PCT * 100}% | TP ${cfg.TAKE_PROFIT_PCT * 100}% | fee ${FEE * 100}%/parte | slippage ${SLIPPAGE * 100}%/parte`);
  console.log('═'.repeat(64));

  const all = await fetchCandles();
  if (all.length < 300) throw new Error(`Doar ${all.length} lumânări — minim 300`);

  let balance = START_BAL, position = null, lastLossIdx = -1e9;
  const trades = [];
  const cooldownBars = Math.ceil(cfg.COOLDOWN_AFTER_LOSS_MIN / 5);
  let equityPeak = START_BAL, maxDD = 0, feesPaid = 0;

  const close = (exitRaw, reason, i) => {
    const dir = position.side === 'BUY' ? 1 : -1;
    const exit = exitRaw * (1 - dir * SLIPPAGE);
    const fee = (position.entryPrice + exit) * position.quantity * FEE;
    const pnl = dir * (exit - position.entryPrice) * position.quantity - fee;
    feesPaid += fee;
    balance += pnl;
    trades.push({ side: position.side, entry: position.entryPrice, exit, pnl, reason, bars: i - position.openedAt });
    if (pnl < 0) lastLossIdx = i;
    position = null;
    equityPeak = Math.max(equityPeak, balance);
    maxDD = Math.max(maxDD, (equityPeak - balance) / equityPeak);
  };

  for (let i = 250; i < all.length; i++) {
    const window = all.slice(0, i + 1);
    const price = all[i].close;

    if (position) {
      const trigger = checkPosition(position, price);
      if (trigger) { close(price, trigger, i); continue; }
    }
    if (position || balance < 1) continue;
    if (i - lastLossIdx < cooldownBars) continue;            // cooldown după pierdere

    const ind = indicators.analyze(window);
    const strat = strategies.analyze(window);
    const sig = criteriaSignal(ind, strat);
    if (sig.action === 'HOLD') continue;
    if (parseFloat(ind.volumeRatio) < parseFloat(process.env.MIN_VOLUME_RATIO || '0.7')) continue;

    // Veto Livermore+Turtle (identic cu live)
    const contra = strat.livermore.trend === 'BEARISH' && strat.turtle.signal === 'SELL' ? 'BUY'
                 : strat.livermore.trend === 'BULLISH' && strat.turtle.signal === 'BUY' ? 'SELL' : null;
    if ((strat.livermore.strength ?? 0) >= 0.8 && sig.action === contra) continue;

    // Filtru HTF 1h (identic cu live). BT_HTF_STRICT: intră DOAR pe direcția
    // trendului mare (NEUTRAL = stai pe mâini), nu doar blochează contra-trend.
    if (cfg.HTF_FILTER) {
      const ratio = TIMEFRAME === '15m' ? 4 : TIMEFRAME === '1h' ? 1 : 12;
      const htf = ratio === 1 ? strategies.htfTrend(window.slice(-80))
                              : strategies.htfTrend(resample1h(window.slice(-60 * ratio), ratio));
      if (process.env.BT_HTF_STRICT === 'true') {
        if (htf !== (sig.action === 'BUY' ? 'BULLISH' : 'BEARISH')) continue;
      } else if ((sig.action === 'BUY' && htf === 'BEARISH') || (sig.action === 'SELL' && htf === 'BULLISH')) {
        continue;
      }
    }

    const mult = strategies.druckenmillerMultiplier(70, sig.criteriaScore, strat.livermore, strat.turtle);
    const dir = sig.action === 'BUY' ? 1 : -1;
    const entry = price * (1 + dir * SLIPPAGE);
    const quantity = (balance * cfg.RISK_PER_TRADE * mult) / entry;
    const { stopLoss, takeProfit } = calcSLTP(sig.action, entry, parseFloat(ind.atr));
    position = { side: sig.action, entryPrice: entry, quantity, stopLoss, takeProfit,
                 initialStop: stopLoss, openedAt: i,
                 trailHigh: sig.action === 'BUY' ? entry : null,
                 trailLow: sig.action === 'SELL' ? entry : null };
  }
  if (position) close(all[all.length - 1].close, 'END', all.length - 1);

  // ─── Raport ────────────────────────────────────────────
  const wins = trades.filter(t => t.pnl > 0), losses = trades.filter(t => t.pnl <= 0);
  const gw = wins.reduce((a, t) => a + t.pnl, 0), gl = Math.abs(losses.reduce((a, t) => a + t.pnl, 0));
  const pf = gl > 0 ? (gw / gl).toFixed(2) : '∞';
  const ret = ((balance - START_BAL) / START_BAL * 100).toFixed(2);
  const barsDays = ((all.length - 250) * (TIMEFRAME === '5m' ? 5 : 15) / 1440).toFixed(1);

  console.log(`\n  Perioadă: ~${barsDays} zile (${all.length - 250} lumânări) | Trades: ${trades.length} (✅ ${wins.length} / ❌ ${losses.length})`);
  console.log(`  Win rate: ${trades.length ? (wins.length / trades.length * 100).toFixed(1) : 0}% | Profit factor: ${pf}`);
  console.log(`  Rezultat net: ${ret >= 0 ? '+' : ''}${ret}% ($${balance.toFixed(2)}) | Comisioane+slippage plătite: $${feesPaid.toFixed(2)}`);
  console.log(`  Max drawdown: -${(maxDD * 100).toFixed(1)}%`);
  if (trades.length) {
    const exp = trades.reduce((a, t) => a + t.pnl, 0) / trades.length;
    console.log(`  Expectancy: ${exp >= 0 ? '+' : ''}$${exp.toFixed(4)}/trade`);
    const byReason = {};
    trades.forEach(t => byReason[t.reason] = (byReason[t.reason] || 0) + 1);
    console.log(`  Exituri: ${Object.entries(byReason).map(([r, n]) => `${r}×${n}`).join(' | ')}`);
  }
  console.log('\n  ⚠️  Stratul AI nu e simulat (filtrează în plus la live). Rezultatele');
  console.log('      trecute nu garantează nimic. Testează pe paper înainte de live.');
  console.log('═'.repeat(64) + '\n');
  return { balance, trades: trades.length, pf };
}

runBacktest().catch(err => { console.error('Backtest error:', err.message); process.exit(1); });
